Rdf4j update module (#172)
kbss-cvut/s-pipes: doc/examples/rdf4j-update/rdf4j-update.md

## Script structure

Script performs following steps:
1) Creates RDF4J repository using `kbss:rdf4j-create-repository` module. Repository is specified by parameters `rdf4j:p-rdf4j-server-url` and `rdf4j:p-rdf4j-repository-name`. Parameter `rdf4j:p-rdf4j-ignore-if-exists` defines behavior of the module in case when defined repository already exists:
in case if it is set to be "true", module will not do anything if repository with given ID already exists on the server.  

    :create-repository
        a kbss:rdf4j-create-repository ;
        rdf4j:p-rdf4j-server-url "http://localhost:8080/rdf4j-server/" ;
        rdf4j:p-rdf4j-repository-name "test-update" ;
        rdf4j:p-rdf4j-ignore-if-exists "true" ;
        sm:next :update ;
    .

2) Perform an update on repository using `kbss:rdf4j-update` module. Repository is defined in the same way as in `:create-repository`. Update query is set by string in `sp:text` section.

    :update-repository
        a kbss:rdf4j-update ;
        sm:next :update-repository_Return ;
        sml:updateQuery [
            a sp:Update ;
            sp:text """
    DELETE {

>         <http://example.org/people/john>  <http://example.org/people/age> ?oldAge .

Reviewer: Done

RDF4J update example explains how to update RDF4J repository.

## Introduction

> This example contains script *rdf4j-update.sms.ttl* that contain one pipeline. It creates RDF4J repository and performs update query. For more details about script construction and execution you can see [hello-world-example](https://github.com/kbss-cvut/s-pipes/blob/main/doc/examples/hello-world/hello-world.md).

Author: ```suggestion
This example explains script [rdf4j-update.sms.ttl](./rdf4j-update.sms.ttl]) that contain one pipeline. It creates RDF4J repository and performs update query. For more details about script construction and execution you can see [hello-world-example](../hello-world/hello-world.md).
```

Author: - provide links whenever possible
- use relative links instead so when we move repository/directory we do not need to refactor

## Script structure

Script performs following steps:
1) Creates RDF4J repository using `kbss:rdf4j-create-repository` module. Repository is specified by parameters `rdf4j:p-rdf4j-server-url` and `rdf4j:p-rdf4j-repository-name`. Parameter `rdf4j:p-rdf4j-ignore-if-exists` defines behavior of the module in case when defined repository already exists:
in case if it is set to be "true", module will not do anything if repository with given ID already exists on the server.  

    :create-repository
        a kbss:rdf4j-create-repository ;
        rdf4j:p-rdf4j-server-url "http://localhost:8080/rdf4j-server/" ;
        rdf4j:p-rdf4j-repository-name "test-update" ;
        rdf4j:p-rdf4j-ignore-if-exists "true" ;
        sm:next :update ;
    .

> 2) Perform an update on repository using `a kbss:rdf4j-update` module. Repository is defined in the same way as in `:create-repository`. Update query is set by string in `sp:text` section.

Author: ```suggestion
2) Perform an update on repository using `kbss:rdf4j-update` module. Repository is defined in the same way as in `:create-repository`. Update query is set by string in `sp:text` section.
```

>     :update

Author: ```suggestion
    :update-repository
```
The previous module you called create-repository, so I would call this update-repository.

Note, that you must update it in the script as well.

Author: In case there is collision in names, use something like:
`:update-repository-default-context`

        a kbss:rdf4j-update ;
        sm:next :update-repository_Return ;
        sml:updateQuery [
            a sp:Update ;
            sp:text """
    DELETE {

>         <http://example.org/people/john>  <http://example.org/people/age> ?oldAge .

Author: Please define a prefix in the script:
`ex-people: http://example.org/people/`

so this si more readable and we demonstrate and test that prefixes are inherited from script file.

Reviewer: There is a problem with defining a prefix:
I have included it in the script like the other prefixes:
`@prefix ex-people: <http://example.org/people/> .`

but I get this error:
```
03-07-2023 18:04:34.413 [http-nio-8080-exec-10] ERROR c.c.s.m.Rdf4jUpdateModule - Update execution exception, query text:

DELETE {
	ex-people:john ex-people:age ?oldAge .
}
INSERT {
	ex-people:john ex-people:age ?newAge .
} WHERE {
   OPTIONAL {
      ex-people:john ex-people:age ?oldAge .
   }
   BIND(COALESCE(?oldAge+1, 1) as ?newAge)
}
      
org.eclipse.rdf4j.query.parser.sparql.ast.VisitorException: QName 'ex-people:john' uses an undefined prefix
```

Author: I suggest following: 
- add prefix inside of sp:text to fix the issue
- make new github issue to add this new feature
- add link of this new issue to the description of this issue (so the issues are interlinked)
- we discuss how to implement it on next meeting (put a note to the agenda please)

Reviewer: Resolved

    }
    INSERT {
        <http://example.org/people/john>  <http://example.org/people/age> ?newAge .
    } WHERE {
        OPTIONAL {
        <http://example.org/people/john>  <http://example.org/people/age> ?oldAge .
        }
        BIND(COALESCE(?oldAge+1, 1) as ?newAge)
    }
            """ ;
        ];
        rdf4j:p-rdf4j-server-url "http://localhost:8080/rdf4j-server/" ;
        rdf4j:p-rdf4j-repository-name "test-update" ;

>         rdfs:label "Make insert update" ;

Author: ```suggestion
```

Author: You do not have label in the previous module ... so I suggest omitting it here as well.

    .

## Script execution

Let's assume that SPipes web application is running at `http://localhost:8080/s-pipes`. We can call the *pipeline* with:

    http://localhost:8080/s-pipes/service?_pId=update-repository

Note, that rdf4j server should be running on the URL that is specified in script (`http://localhost:8080/rdf4j-server/` for example).
You can see following logs while execution:

    [http-nio-8080-exec-7] INFO  c.c.s.e.ExecutionEngineImpl -  ##### create-repository
    ...

>     INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jCreateRepositoryModule.executeSelf Server url: http://localhost:8080/rdf4j-server/ Repository name: test-update Ignore flag: true

Author: ```suggestion
    INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jCreateRepositoryModule.executeSelf Server url: http://localhost:8080/rdf4j-server/, Repository name: test-update, Ignore if repository exists: true
```

Author: Note it should be fixed in code as well.

    INFO [http-nio-8080-exec-6] cz.cvut.spipes.modules.Rdf4jCreateRepositoryModule.executeSelf Repository "test-update" already exists
    ...
    [http-nio-8080-exec-7] INFO  c.c.s.e.ExecutionEngineImpl -  ##### Make insert update
    ...

>     INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.loadConfiguration Connection created for repository test-update

Author: ```suggestion
    DEBUG [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.loadConfiguration Connected to repository test-update
```

>     INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.makeUpdate Update executed

Author: ```suggestion
    DEBUG [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.makeUpdate Update successful
```
